### PR TITLE
Add overlay.py to allow for overlay architecture generation

### PR DESCRIPTION
**Pull request by andrewb1999**
Signed-off-by: Andrew Butt <butta@seas.upenn.edu>

Essentially an inverse roi.  Can accommodate cutting out multiple roi like regions.  Testing with symbiflow-arch-defs overlay generation and working.

### CODE REVIEW

File: prjxray/overlay.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+class Overlay(object):
+    """ Object that represents an overlay.
+
+    Can be used to iterate over tiles and sites not inside a partition region.
+
+    """
+
+    def __init__(self, db, region_dict):
```
Comment (litghost) [inline prjxray/overlay.py:8]: Why not just take the grid?

File: prjxray/overlay.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+class Overlay(object):
```
Comment (litghost) [inline prjxray/overlay.py:1]: Tests?

File: prjxray/overlay.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                continue
+
+            gridinfo = self.grid.gridinfo_at_loc(loc)
+
+            if tile_types is not None and gridinfo.tile_type not in tile_types:
+                continue
+
+            yield tile_name
+
+    def gen_sites(self, site_types=None):
```
Comment (litghost) [inline prjxray/overlay.py:41]: This seems like a less useful method, as it is just an extension of `gen_tiles`.  Why have both?

File: prjxray/overlay.py (hunk 1 above)
Comment (andrewb1999) [reply prjxray/overlay.py:8]: Which grid are you talking about?  I'm trying to match the original roi input schema which was x_max, x_min, y_min, y_max.  Just abstracting that to take multiple regions.

File: prjxray/overlay.py (hunk 2 above)
Comment (andrewb1999) [reply prjxray/overlay.py:1]: Not sure what a proper test for this in prjxray would be.  Any suggestions?

File: prjxray/overlay.py (hunk 3 above)
Comment (andrewb1999) [reply prjxray/overlay.py:41]: Actually, I was just leaving in both the gen_tiles and gen_sites methods to match the roi interface.  I do not believe they are used in architecture generation.  Maybe I should just get rid of both methods?

Comment (litghost) [reply prjxray/overlay.py:41]: Drop both

File: prjxray/overlay.py (hunk 2 above)
Comment (litghost) [reply prjxray/overlay.py:1]: Provide a grid and a region dict and ensure that things work as expected?

File: prjxray/overlay.py (hunk 1 above)
Comment (litghost) [reply prjxray/overlay.py:8]: You take the `db` argument, but immediately grab the grid, and discard the db.  Just take the grid argument.

File: prjxray/overlay.py (hunk 2 above)
Comment (andrewb1999) [reply prjxray/overlay.py:1]: Ok makes sense.  Should that be a new minitests subdirectory?

Comment (litghost) [reply prjxray/overlay.py:1]: We have https://github.com/SymbiFlow/prjxray/tree/master/tests and I believe we are running them via pytest.

Comment (litghost) [reply prjxray/overlay.py:1]: https://github.com/SymbiFlow/prjxray/blob/master/Makefile#L59-L60

**Comment by andrewb1999** [comment]
@litghost This test case works on my local machine, but I guess it doesn't work here because the database isn't built.  What would you suggest doing?

File: prjxray/overlay.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+# SPDX-License-Identifier: ISC
+class Overlay(object):
+    """ Object that represents an overlay.
+
+    Can be used to iterate over tiles and sites not inside a partition region.
+
+    """
+
+    def __init__(self, grid, region_dict):
+        self.grid = grid
```
Comment (litghost) [inline prjxray/overlay.py:19]: Is `self.grid` even used anymore?

Comment (andrewb1999) [reply prjxray/overlay.py:19]: Oh you're right, it's not.  That fixes the testing issue as well.

**Comment by andrewb1999** [comment]
@litghost can this be merged?

**Pull request merged**